# Creating the Dataset

In [92]:
import tabula
import pandas as pd
import numpy as np
import re

In [4]:
hanja_file = 'hanja.pdf'
list_hanja = tabula.read_pdf(hanja_file, pages = "all", multiple_tables = True)
list_hanja

[      0                                            1
 0   NaN                     한문 교육용 인명용 추가 한자 및 허용 한자
 1    한글                                         기초한자
 2   NaN                          (2007.8.현재) 별표1 별표2
 3   NaN                 家佳街可歌 嘉嫁稼賈駕伽迦柯呵哥枷珂痂苛茄袈訶跏軻哿嘏舸
 4     가                                          NaN
 5   NaN                                      加價假架暇 珈
 6   NaN                                        各角脚閣却
 7     각                                    珏恪殼愨 愨(慤)
 8   NaN                                           覺刻
 9   NaN                                        干間看刊肝
 10    간                 艮侃杆玕竿揀諫墾栞奸柬澗磵稈艱癇忓矸 杆(桿) 癇(癎)
 11  NaN                                         幹簡姦懇
 12    갈                                  渴 葛乫喝曷碣竭褐蝎鞨
 13  NaN                                        甘減感敢監
 14    감                          勘堪瞰坎嵌憾戡柑橄疳紺邯龕玪 鑑(鑒)
 15  NaN                                            鑑
 16    갑                                      甲 鉀匣岬胛閘
 17  NaN  江降講強康 杠堈岡姜橿彊慷畺疆糠絳羌

In [5]:
len(list_hanja)

53

In [6]:
type(list_hanja)

list

In [7]:
list_hanja[0]

0                                            1
0   NaN                     한문 교육용 인명용 추가 한자 및 허용 한자
1    한글                                         기초한자
2   NaN                          (2007.8.현재) 별표1 별표2
3   NaN                 家佳街可歌 嘉嫁稼賈駕伽迦柯呵哥枷珂痂苛茄袈訶跏軻哿嘏舸
4     가                                          NaN
5   NaN                                      加價假架暇 珈
6   NaN                                        各角脚閣却
7     각                                    珏恪殼愨 愨(慤)
8   NaN                                           覺刻
9   NaN                                        干間看刊肝
10    간                 艮侃杆玕竿揀諫墾栞奸柬澗磵稈艱癇忓矸 杆(桿) 癇(癎)
11  NaN                                         幹簡姦懇
12    갈                                  渴 葛乫喝曷碣竭褐蝎鞨
13  NaN                                        甘減感敢監
14    감                          勘堪瞰坎嵌憾戡柑橄疳紺邯龕玪 鑑(鑒)
15  NaN                                            鑑
16    갑                                      甲 鉀匣岬胛閘
17  NaN  江降講強康 杠堈岡姜橿彊慷畺疆糠絳羌腔舡薑鱇嫝跭 玒顜茳 強(强) 鋼(鎠) 岡(崗)
18    강                                          NaN
19  NaN                                    剛鋼綱 鏹 (襁)
20  NaN                                        改皆個開介
21    개                        价凱愷漑塏愾疥芥豈鎧玠 個(箇) 蓋(盖)
22  NaN                                          慨槪蓋

In [8]:
type(list_hanja[0])

pandas.core.frame.DataFrame

The list_hanja is a list made up of 53 dataframes, we need to combine these and clean it up.

In [11]:
df_hanja = pd.DataFrame()

In [12]:
for i in range(len(list_hanja)):
    df_hanja = df_hanja.append(list_hanja[i])
df_hanja

0                                            1  \
0   NaN                     한문 교육용 인명용 추가 한자 및 허용 한자   
1    한글                                         기초한자   
2   NaN                          (2007.8.현재) 별표1 별표2   
3   NaN                 家佳街可歌 嘉嫁稼賈駕伽迦柯呵哥枷珂痂苛茄袈訶跏軻哿嘏舸   
4     가                                          NaN   
5   NaN                                      加價假架暇 珈   
6   NaN                                        各角脚閣却   
7     각                                    珏恪殼愨 愨(慤)   
8   NaN                                           覺刻   
9   NaN                                        干間看刊肝   
10    간                 艮侃杆玕竿揀諫墾栞奸柬澗磵稈艱癇忓矸 杆(桿) 癇(癎)   
11  NaN                                         幹簡姦懇   
12    갈                                  渴 葛乫喝曷碣竭褐蝎鞨   
13  NaN                                        甘減感敢監   
14    감                          勘堪瞰坎嵌憾戡柑橄疳紺邯龕玪 鑑(鑒)   
15  NaN                                            鑑   
16    갑                                      甲 鉀匣岬胛閘   
17  NaN  江降講強康 杠堈岡姜橿彊慷畺疆糠絳羌腔舡薑鱇嫝跭 玒顜茳 強(强) 鋼(鎠) 岡(崗)   
18    강                                          NaN   
19  NaN                                    剛鋼綱 鏹 (襁)   
20  NaN                                        改皆個開介   
21    개                        价凱愷漑塏愾疥芥豈鎧玠 個(箇) 蓋(盖)   
22  NaN                                          慨槪蓋   
0    한글                    한문 교육용\r기초한자\r(2007.8.현재)   
1   NaN                                          NaN   
2     객                                            客   
3     갱                                            更   
4     갹                                          NaN   
5     거                                   去巨居車擧\r距拒據   
6     건                                         建乾件健   
..  ...                                          ...   
9     환                                     歡患丸換環\r還   
0    한글                    한문 교육용\r기초한자\r(2007.8.현재)   
1   NaN                                          NaN   
2     활                                            活   
3     황                                         黃皇況荒   
4     회                                         回會悔懷   
5     획                                           獲劃   
6     횡                                            橫   
7     효                                          孝效曉   
8     후                                         後厚侯候   
9     훈                                            訓   
10    훙                                          NaN   
0    한글                    한문 교육용\r기초한자\r(2007.8.현재)   
1   NaN                                          NaN   
2     훤                                          NaN   
3     훼                                            毁   
4     휘                                           揮輝   
5     휴                                           休携   
6     휼                                          NaN   
7     흉                                           凶胸   
8     흑                                            黑   
9     흔                                          NaN   
10    흘                                          NaN   
11    흠                                          NaN   
12    흡                                            吸   
0    한글                    한문 교육용\r기초한자\r(2007.8.현재)   
1   NaN                                          NaN   
2     흥                                            興   
3     희                                         希喜稀戲   
4     힐                                          NaN   

                            2                         3    4    5  
0                         NaN                       NaN  NaN  NaN  
1                         NaN                       NaN  NaN  NaN  
2                         NaN                       NaN  NaN  NaN  
3                         NaN                       NaN  NaN  NaN  
4                         NaN                       NaN  NaN  NaN  
5                         NaN                       NaN  NaN  NaN  
6                         NaN                       NaN  

There's a lot of superfluous data, so we will remove them.

In [13]:
df_hanja = df_hanja[df_hanja[0] != '한글']
df_hanja = df_hanja[df_hanja[2] != '별표1']
df_hanja = df_hanja[2:]
df_hanja

0                                            1  \
3   NaN                 家佳街可歌 嘉嫁稼賈駕伽迦柯呵哥枷珂痂苛茄袈訶跏軻哿嘏舸   
4     가                                          NaN   
5   NaN                                      加價假架暇 珈   
6   NaN                                        各角脚閣却   
7     각                                    珏恪殼愨 愨(慤)   
8   NaN                                           覺刻   
9   NaN                                        干間看刊肝   
10    간                 艮侃杆玕竿揀諫墾栞奸柬澗磵稈艱癇忓矸 杆(桿) 癇(癎)   
11  NaN                                         幹簡姦懇   
12    갈                                  渴 葛乫喝曷碣竭褐蝎鞨   
13  NaN                                        甘減感敢監   
14    감                          勘堪瞰坎嵌憾戡柑橄疳紺邯龕玪 鑑(鑒)   
15  NaN                                            鑑   
16    갑                                      甲 鉀匣岬胛閘   
17  NaN  江降講強康 杠堈岡姜橿彊慷畺疆糠絳羌腔舡薑鱇嫝跭 玒顜茳 強(强) 鋼(鎠) 岡(崗)   
18    강                                          NaN   
19  NaN                                    剛鋼綱 鏹 (襁)   
20  NaN                                        改皆個開介   
21    개                        价凱愷漑塏愾疥芥豈鎧玠 個(箇) 蓋(盖)   
22  NaN                                          慨槪蓋   
2     객                                            客   
3     갱                                            更   
4     갹                                          NaN   
5     거                                   去巨居車擧\r距拒據   
6     건                                         建乾件健   
7     걸                                           傑乞   
8     검                                          儉劍檢   
9     겁                                          NaN   
10    게                                          NaN   
11    격                                         格擊激隔   
..  ...                                          ...   
3     혹                                           或惑   
4     혼                                         婚混昏魂   
5     홀                                            忽   
6     홍                                         紅洪弘鴻   
7     화                                 火化花貨和\r話畫華禾禍   
8     확                                          確穫擴   
9     환                                     歡患丸換環\r還   
2     활                                            活   
3     황                                         黃皇況荒   
4     회                                         回會悔懷   
5     획                                           獲劃   
6     횡                                            橫   
7     효                                          孝效曉   
8     후                                         後厚侯候   
9     훈                                            訓   
10    훙                                          NaN   
2     훤                                          NaN   
3     훼                                            毁   
4     휘                                           揮輝   
5     휴                                           休携   
6     휼                                          NaN   
7     흉                                           凶胸   
8     흑                                            黑   
9     흔                                          NaN   
10    흘                                          NaN   
11    흠                                          NaN   
12    흡                                            吸   
2     흥                                            興   
3     희                                         希喜稀戲   
4     힐                                          NaN   

                            2                         3    4    5  
3                         NaN                       NaN  NaN  NaN  
4                         NaN                       NaN  NaN  NaN  
5                         NaN                       NaN  NaN  NaN  
6                         NaN                       NaN  NaN  NaN  
7                         NaN                       NaN  NaN  NaN  
8                         NaN                       NaN  NaN  NaN  
9                         NaN                       NaN  

In [14]:
df_hanja = df_hanja.rename(columns={0:'hangul', 1:'hanja', 2:'hanja2', 3:'hanja3', 4:'hanja4', 5:'hanja5'})

In [15]:
df_hanja.head(25)

hangul                                        hanja       hanja2  \
3     NaN                 家佳街可歌 嘉嫁稼賈駕伽迦柯呵哥枷珂痂苛茄袈訶跏軻哿嘏舸          NaN   
4       가                                          NaN          NaN   
5     NaN                                      加價假架暇 珈          NaN   
6     NaN                                        各角脚閣却          NaN   
7       각                                    珏恪殼愨 愨(慤)          NaN   
8     NaN                                           覺刻          NaN   
9     NaN                                        干間看刊肝          NaN   
10      간                 艮侃杆玕竿揀諫墾栞奸柬澗磵稈艱癇忓矸 杆(桿) 癇(癎)          NaN   
11    NaN                                         幹簡姦懇          NaN   
12      갈                                  渴 葛乫喝曷碣竭褐蝎鞨          NaN   
13    NaN                                        甘減感敢監          NaN   
14      감                          勘堪瞰坎嵌憾戡柑橄疳紺邯龕玪 鑑(鑒)          NaN   
15    NaN                                            鑑          NaN   
16      갑                                      甲 鉀匣岬胛閘          NaN   
17    NaN  江降講強康 杠堈岡姜橿彊慷畺疆糠絳羌腔舡薑鱇嫝跭 玒顜茳 強(强) 鋼(鎠) 岡(崗)          NaN   
18      강                                          NaN          NaN   
19    NaN                                    剛鋼綱 鏹 (襁)          NaN   
20    NaN                                        改皆個開介          NaN   
21      개                        价凱愷漑塏愾疥芥豈鎧玠 個(箇) 蓋(盖)          NaN   
22    NaN                                          慨槪蓋          NaN   
2       객                                            客            喀   
3       갱                                            更          坑粳羹   
4       갹                                          NaN            醵   
5       거                                   去巨居車擧\r距拒據    渠遽鉅炬倨据祛踞鋸   
6       건                                         建乾件健  巾虔楗鍵愆腱蹇騫搴湕踺   

       hanja3 hanja4 hanja5  
3         NaN    NaN    NaN  
4         NaN    NaN    NaN  
5         NaN    NaN    NaN  
6         NaN    NaN    NaN  
7         NaN    NaN    NaN  
8         NaN    NaN    NaN  
9         NaN    NaN    NaN  
10        NaN    NaN    NaN  
11        NaN    NaN    NaN  
12        NaN    NaN    NaN  
13        NaN    NaN    NaN  
14        NaN    NaN    NaN  
15        NaN    NaN    NaN  
16        NaN    NaN    NaN  
17        NaN    NaN    NaN  
18        NaN    NaN    NaN  
19        NaN    NaN    NaN  
20        NaN    NaN    NaN  
21        NaN    NaN    NaN  
22        NaN    NaN    NaN  
2         NaN    NaN    NaN  
3         NaN    NaN    NaN  
4         NaN    NaN    NaN  
5         NaN    NaN    NaN  
6   建(䢖) 乾(漧)    NaN    NaN

The Hanja is out of step with the corresponding Hangul, we will fix this next. It seems part of the Hanja appears above and below the Hangul resulting in a NaN. We have to do this manually because of the inconsistency

In [16]:
df_hanja.iat[1,1] = str(df_hanja.iat[0,1]) + str(df_hanja.iat[2,1])
df_hanja.iat[4,1] += str(df_hanja.iat[3,1]) + str(df_hanja.iat[5,1])
df_hanja.iat[7,1] += str(df_hanja.iat[6,1]) + str(df_hanja.iat[8,1])
df_hanja.iat[11,1] += str(df_hanja.iat[10,1]) + str(df_hanja.iat[12,1])
df_hanja.iat[15,1] = str(df_hanja.iat[14,1]) + str(df_hanja.iat[16,1])
df_hanja.iat[18,1] += str(df_hanja.iat[17,1]) + str(df_hanja.iat[19,1])
df_hanja

hangul                                              hanja  \
3     NaN                       家佳街可歌 嘉嫁稼賈駕伽迦柯呵哥枷珂痂苛茄袈訶跏軻哿嘏舸   
4       가                家佳街可歌 嘉嫁稼賈駕伽迦柯呵哥枷珂痂苛茄袈訶跏軻哿嘏舸加價假架暇 珈   
5     NaN                                            加價假架暇 珈   
6     NaN                                              各角脚閣却   
7       각                                   珏恪殼愨 愨(慤)各角脚閣却覺刻   
8     NaN                                                 覺刻   
9     NaN                                              干間看刊肝   
10      간              艮侃杆玕竿揀諫墾栞奸柬澗磵稈艱癇忓矸 杆(桿) 癇(癎)干間看刊肝幹簡姦懇   
11    NaN                                               幹簡姦懇   
12      갈                                        渴 葛乫喝曷碣竭褐蝎鞨   
13    NaN                                              甘減感敢監   
14      감                          勘堪瞰坎嵌憾戡柑橄疳紺邯龕玪 鑑(鑒)甘減感敢監鑑   
15    NaN                                                  鑑   
16      갑                                            甲 鉀匣岬胛閘   
17    NaN        江降講強康 杠堈岡姜橿彊慷畺疆糠絳羌腔舡薑鱇嫝跭 玒顜茳 強(强) 鋼(鎠) 岡(崗)   
18      강  江降講強康 杠堈岡姜橿彊慷畺疆糠絳羌腔舡薑鱇嫝跭 玒顜茳 強(强) 鋼(鎠) 岡(崗)剛鋼綱...   
19    NaN                                          剛鋼綱 鏹 (襁)   
20    NaN                                              改皆個開介   
21      개                      价凱愷漑塏愾疥芥豈鎧玠 個(箇) 蓋(盖)改皆個開介慨槪蓋   
22    NaN                                                慨槪蓋   
2       객                                                  客   
3       갱                                                  更   
4       갹                                                NaN   
5       거                                         去巨居車擧\r距拒據   
6       건                                               建乾件健   
7       걸                                                 傑乞   
8       검                                                儉劍檢   
9       겁                                                NaN   
10      게                                                NaN   
11      격                                               格擊激隔   
..    ...                                                ...   
3       혹                                                 或惑   
4       혼                                               婚混昏魂   
5       홀                                                  忽   
6       홍                                               紅洪弘鴻   
7       화                                       火化花貨和\r話畫華禾禍   
8       확                                                確穫擴   
9       환                                           歡患丸換環\r還   
2       활                                                  活   
3       황                                               黃皇況荒   
4       회                                               回會悔懷   
5       획                                                 獲劃   
6       횡                                                  橫   
7       효                                                孝效曉   
8       후                                               後厚侯候   
9       훈                                                  訓   
10      훙                                                NaN   
2       훤                                                NaN   
3       훼                                                  毁   
4       휘                                                 揮輝   
5       휴                                                 休携   
6       휼                                                NaN   
7       흉                                                 凶胸   
8       흑                                                  黑   
9       흔                                                NaN   
10      흘                                                NaN   
11      흠                                                NaN   
12      흡                                                  吸   
2       흥                                                  興   
3       희                                               希喜稀戲   
4       힐                                                NaN   

                       hanja2     

Combine all hanja columns

In [17]:
df_hanja = df_hanja[pd.notnull(df_hanja['hangul'])]
df_hanja = df_hanja.replace(np.nan, '', regex=True)
df_hanja['hanja'] += df_hanja['hanja2'] + df_hanja['hanja3'] + df_hanja['hanja4'] + df_hanja['hanja5']
del df_hanja['hanja2'], df_hanja['hanja3'], df_hanja['hanja4'], df_hanja['hanja5']
df_hanja

hangul                                              hanja
4       가                家佳街可歌 嘉嫁稼賈駕伽迦柯呵哥枷珂痂苛茄袈訶跏軻哿嘏舸加價假架暇 珈
7       각                                   珏恪殼愨 愨(慤)各角脚閣却覺刻
10      간              艮侃杆玕竿揀諫墾栞奸柬澗磵稈艱癇忓矸 杆(桿) 癇(癎)干間看刊肝幹簡姦懇
12      갈                                        渴 葛乫喝曷碣竭褐蝎鞨
14      감                          勘堪瞰坎嵌憾戡柑橄疳紺邯龕玪 鑑(鑒)甘減感敢監鑑
16      갑                                            甲 鉀匣岬胛閘
18      강  江降講強康 杠堈岡姜橿彊慷畺疆糠絳羌腔舡薑鱇嫝跭 玒顜茳 強(强) 鋼(鎠) 岡(崗)剛鋼綱...
21      개                      价凱愷漑塏愾疥芥豈鎧玠 個(箇) 蓋(盖)改皆個開介慨槪蓋
2       객                                                 客喀
3       갱                                               更坑粳羹
4       갹                                                  醵
5       거                                去巨居車擧\r距拒據渠遽鉅炬倨据祛踞鋸
6       건                           建乾件健巾虔楗鍵愆腱蹇騫搴湕踺建(䢖) 乾(漧)
7       걸                                            傑乞桀傑(杰)
8       검                                         儉劍檢瞼鈐黔劍(劒)
9       겁                                                劫怯迲
10      게                                                揭偈憩
11      격                                            格擊激隔檄膈覡
12      견                                      犬見堅肩絹\r遣牽鵑甄繭譴
13      결                                    決結潔缺訣抉㛃焆迼玦鍥潔(㓗)
2       경  京景經庚耕\r敬輕驚慶競\r竟境鏡頃傾\r硬警徑卿倞鯨坰耿炅更梗憬璟瓊擎儆俓涇莖勁逕熲冏勍\...
3       계             癸季界計溪\r鷄系係戒械\r繼契桂啓階\r繫誡烓屆悸棨稽谿界(堺) 谿(磎)
4       고  古故固苦高\r考告枯姑庫\r孤鼓稿顧叩敲皐暠呱尻拷槁沽痼睾羔股膏苽菰藁蠱袴誥\r賈辜錮雇杲鼔...
5       곡                                            谷曲穀哭斛梏鵠
6       곤                                    困坤昆崑琨錕梱棍滾鯤衮衮(袞)
7       골                                                骨汨滑
2       공                              工功空共公\r孔供恭攻恐\r貢珙控拱蚣鞏龔
3       곶                                                  串
4       과                                     果課科過誇\r寡菓跨鍋顆戈瓜
5       곽                                               郭廓槨藿
..    ...                                                ...
3       혹                                               或惑酷熇
4       혼                                           婚混昏魂渾琿俒䫟
5       홀                                                忽惚笏
6       홍                                        紅洪弘鴻泓烘虹鉷哄汞訌
7       화                              火化花貨和\r話畫華禾禍嬅樺譁靴澕畫(畵)
8       확                                          確穫擴廓攫確(碻)
9       환                            歡患丸換環\r還喚奐渙煥晥幻桓鐶驩宦紈鰥圜皖洹
2       활                                          活闊滑猾豁闊(濶)
3       황                  黃皇況荒凰堭媓晃滉榥煌璜熀幌徨恍惶愰慌湟潢篁簧蝗\r遑隍楻晃(晄)
4       회                      回會悔懷廻恢晦檜澮繪誨匯徊淮獪膾茴蛔賄灰繪(絵) 會(会)
5       획                                                獲劃画
6       횡                                                橫鐄宖
7       효                       孝效曉涍爻驍斅哮嚆梟淆肴酵皛歊窙謼傚洨庨虓熇烋婋效(効)
8       후                              後厚侯候后逅吼嗅帿朽煦珝喉堠㰭姁芋厚(垕)
9       훈                    訓勳焄熏薰壎燻鑂暈纁煇薫勳(勛ᆞ) 薰() 壎(塤)\r熏()
10      훙                                                  薨
2       훤                                             喧暄萱煊愃昍
3       훼                                              毁喙(卉)
4       휘                                          揮輝彙徽暉煇諱麾煒
5       휴                                             休携烋畦虧庥
6       휼                                                恤譎鷸
7       흉                                              凶胸兇匈洶
8       흑                                                  黑
9       흔                                              欣炘昕痕忻
10      흘                                               屹吃紇訖
11      흠                                               欽欠歆鑫
12      흡                                               吸洽恰翕
2       흥                                                  興
3       희  希喜稀戲姬晞僖橲禧嬉憙熹 羲爔曦俙 憘犧噫熙烯暿譆\r㜯煕(熙 . 熈) 熹(熺) 戲(戱)...
4       힐                                                  詰

[453 rows x 2 columns]

Finally we need to reset the index

In [18]:
df_hanja = df_hanja.reset_index()
df_hanja = df_hanja.iloc[:,1:3]
df_hanja

hangul                                              hanja
0        가                家佳街可歌 嘉嫁稼賈駕伽迦柯呵哥枷珂痂苛茄袈訶跏軻哿嘏舸加價假架暇 珈
1        각                                   珏恪殼愨 愨(慤)各角脚閣却覺刻
2        간              艮侃杆玕竿揀諫墾栞奸柬澗磵稈艱癇忓矸 杆(桿) 癇(癎)干間看刊肝幹簡姦懇
3        갈                                        渴 葛乫喝曷碣竭褐蝎鞨
4        감                          勘堪瞰坎嵌憾戡柑橄疳紺邯龕玪 鑑(鑒)甘減感敢監鑑
5        갑                                            甲 鉀匣岬胛閘
6        강  江降講強康 杠堈岡姜橿彊慷畺疆糠絳羌腔舡薑鱇嫝跭 玒顜茳 強(强) 鋼(鎠) 岡(崗)剛鋼綱...
7        개                      价凱愷漑塏愾疥芥豈鎧玠 個(箇) 蓋(盖)改皆個開介慨槪蓋
8        객                                                 客喀
9        갱                                               更坑粳羹
10       갹                                                  醵
11       거                                去巨居車擧\r距拒據渠遽鉅炬倨据祛踞鋸
12       건                           建乾件健巾虔楗鍵愆腱蹇騫搴湕踺建(䢖) 乾(漧)
13       걸                                            傑乞桀傑(杰)
14       검                                         儉劍檢瞼鈐黔劍(劒)
15       겁                                                劫怯迲
16       게                                                揭偈憩
17       격                                            格擊激隔檄膈覡
18       견                                      犬見堅肩絹\r遣牽鵑甄繭譴
19       결                                    決結潔缺訣抉㛃焆迼玦鍥潔(㓗)
20       경  京景經庚耕\r敬輕驚慶競\r竟境鏡頃傾\r硬警徑卿倞鯨坰耿炅更梗憬璟瓊擎儆俓涇莖勁逕熲冏勍\...
21       계             癸季界計溪\r鷄系係戒械\r繼契桂啓階\r繫誡烓屆悸棨稽谿界(堺) 谿(磎)
22       고  古故固苦高\r考告枯姑庫\r孤鼓稿顧叩敲皐暠呱尻拷槁沽痼睾羔股膏苽菰藁蠱袴誥\r賈辜錮雇杲鼔...
23       곡                                            谷曲穀哭斛梏鵠
24       곤                                    困坤昆崑琨錕梱棍滾鯤衮衮(袞)
25       골                                                骨汨滑
26       공                              工功空共公\r孔供恭攻恐\r貢珙控拱蚣鞏龔
27       곶                                                  串
28       과                                     果課科過誇\r寡菓跨鍋顆戈瓜
29       곽                                               郭廓槨藿
..     ...                                                ...
423      혹                                               或惑酷熇
424      혼                                           婚混昏魂渾琿俒䫟
425      홀                                                忽惚笏
426      홍                                        紅洪弘鴻泓烘虹鉷哄汞訌
427      화                              火化花貨和\r話畫華禾禍嬅樺譁靴澕畫(畵)
428      확                                          確穫擴廓攫確(碻)
429      환                            歡患丸換環\r還喚奐渙煥晥幻桓鐶驩宦紈鰥圜皖洹
430      활                                          活闊滑猾豁闊(濶)
431      황                  黃皇況荒凰堭媓晃滉榥煌璜熀幌徨恍惶愰慌湟潢篁簧蝗\r遑隍楻晃(晄)
432      회                      回會悔懷廻恢晦檜澮繪誨匯徊淮獪膾茴蛔賄灰繪(絵) 會(会)
433      획                                                獲劃画
434      횡                                                橫鐄宖
435      효                       孝效曉涍爻驍斅哮嚆梟淆肴酵皛歊窙謼傚洨庨虓熇烋婋效(効)
436      후                              後厚侯候后逅吼嗅帿朽煦珝喉堠㰭姁芋厚(垕)
437      훈                    訓勳焄熏薰壎燻鑂暈纁煇薫勳(勛ᆞ) 薰() 壎(塤)\r熏()
438      훙                                                  薨
439      훤                                             喧暄萱煊愃昍
440      훼                                              毁喙(卉)
441      휘                                          揮輝彙徽暉煇諱麾煒
442      휴                                             休携烋畦虧庥
443      휼                                                恤譎鷸
444      흉                                              凶胸兇匈洶
445      흑                                                  黑
446      흔                                              欣炘昕痕忻
447      흘                                               屹吃紇訖
448      흠                                               欽欠歆鑫
449      흡                                               吸洽恰翕
450      흥                                                  興
451      희  希喜稀戲姬晞僖橲禧嬉憙熹 羲爔曦俙 憘犧噫熙烯暿譆\r㜯煕(熙 . 熈) 熹(熺) 戲(戱)...
452      힐                                                  詰

[453 rows x 2 columns]

In [22]:
len(df_hanja)

453

We have 453 name parts, from which we can construct over 200,000 names, this is too computationally expensive for testing purposes. We will assume that the number of hanja correlates with how often the syllable appears in names.

Get rid of brackets around characters

In [35]:
df_hanja = df_hanja.replace('\(', '', regex=True)
df_hanja = df_hanja.replace('\)', '', regex=True)

In [38]:
df_hanja['hanja_count'] = df_hanja['hanja'].apply(len)
df_hanja

hangul                                              hanja  hanja_count
0        가                家佳街可歌 嘉嫁稼賈駕伽迦柯呵哥枷珂痂苛茄袈訶跏軻哿嘏舸加價假架暇 珈           35
1        각                                     珏恪殼愨 愨慤各角脚閣却覺刻           14
2        간                  艮侃杆玕竿揀諫墾栞奸柬澗磵稈艱癇忓矸 杆桿 癇癎干間看刊肝幹簡姦懇           33
3        갈                                        渴 葛乫喝曷碣竭褐蝎鞨           11
4        감                            勘堪瞰坎嵌憾戡柑橄疳紺邯龕玪 鑑鑒甘減感敢監鑑           23
5        갑                                            甲 鉀匣岬胛閘            7
6        강       江降講強康 杠堈岡姜橿彊慷畺疆糠絳羌腔舡薑鱇嫝跭 玒顜茳 強强 鋼鎠 岡崗剛鋼綱 鏹 襁           44
7        개                          价凱愷漑塏愾疥芥豈鎧玠 個箇 蓋盖改皆個開介慨槪蓋           25
8        객                                                 客喀            2
9        갱                                               更坑粳羹            4
10       갹                                                  醵            1
11       거                                去巨居車擧\r距拒據渠遽鉅炬倨据祛踞鋸           18
12       건                               建乾件健巾虔楗鍵愆腱蹇騫搴湕踺建䢖 乾漧           20
13       걸                                              傑乞桀傑杰            5
14       검                                           儉劍檢瞼鈐黔劍劒            8
15       겁                                                劫怯迲            3
16       게                                                揭偈憩            3
17       격                                            格擊激隔檄膈覡            7
18       견                                      犬見堅肩絹\r遣牽鵑甄繭譴           12
19       결                                      決結潔缺訣抉㛃焆迼玦鍥潔㓗           13
20       경  京景經庚耕\r敬輕驚慶競\r竟境鏡頃傾\r硬警徑卿倞鯨坰耿炅更梗憬璟瓊擎儆俓涇莖勁逕熲冏勍\...           74
21       계                 癸季界計溪\r鷄系係戒械\r繼契桂啓階\r繫誡烓屆悸棨稽谿界堺 谿磎           31
22       고  古故固苦高\r考告枯姑庫\r孤鼓稿顧叩敲皐暠呱尻拷槁沽痼睾羔股膏苽菰藁蠱袴誥\r賈辜錮雇杲鼔...           48
23       곡                                            谷曲穀哭斛梏鵠            7
24       곤                                      困坤昆崑琨錕梱棍滾鯤衮衮袞           13
25       골                                                骨汨滑            3
26       공                              工功空共公\r孔供恭攻恐\r貢珙控拱蚣鞏龔           19
27       곶                                                  串            1
28       과                                     果課科過誇\r寡菓跨鍋顆戈瓜           13
29       곽                                               郭廓槨藿            4
..     ...                                                ...          ...
423      혹                                               或惑酷熇            4
424      혼                                           婚混昏魂渾琿俒䫟            8
425      홀                                                忽惚笏            3
426      홍                                        紅洪弘鴻泓烘虹鉷哄汞訌           11
427      화                                火化花貨和\r話畫華禾禍嬅樺譁靴澕畫畵           18
428      확                                            確穫擴廓攫確碻            7
429      환                            歡患丸換環\r還喚奐渙煥晥幻桓鐶驩宦紈鰥圜皖洹           22
430      활                                            活闊滑猾豁闊濶            7
431      황                    黃皇況荒凰堭媓晃滉榥煌璜熀幌徨恍惶愰慌湟潢篁簧蝗\r遑隍楻晃晄           30
432      회                          回會悔懷廻恢晦檜澮繪誨匯徊淮獪膾茴蛔賄灰繪絵 會会           25
433      획                                                獲劃画            3
434      횡                                                橫鐄宖            3
435      효                         孝效曉涍爻驍斅哮嚆梟淆肴酵皛歊窙謼傚洨庨虓熇烋婋效効           26
436      후                                後厚侯候后逅吼嗅帿朽煦珝喉堠㰭姁芋厚垕           19
437      훈                            訓勳焄熏薰壎燻鑂暈纁煇薫勳勛ᆞ 薰 壎塤\r熏           22
438      훙                                                  薨            1
439      훤                                             喧暄萱煊愃昍            6
440      훼                                                毁喙卉            3
441      휘                                          揮輝彙徽暉煇諱麾煒            9
442      휴                                             休携烋畦虧庥            6
443      휼                                                恤譎鷸            3
444      흉                   

In [40]:
df_hanja['hanja_count'].describe()

count    453.000000
mean      12.997792
std       14.257974
min        1.000000
25%        4.000000
50%        7.000000
75%       18.000000
max       92.000000
Name: hanja_count, dtype: float64

Let's take the top 25th percentile.

In [47]:
df_subset = pd.DataFrame()

In [48]:
df_subset = df_hanja[df_hanja['hanja_count'] >= 18]
df_subset = df_subset.reset_index()
df_subset = df_subset.iloc[:,1:4] 
df_subset

hangul                                              hanja  hanja_count
0        가                家佳街可歌 嘉嫁稼賈駕伽迦柯呵哥枷珂痂苛茄袈訶跏軻哿嘏舸加價假架暇 珈           35
1        간                  艮侃杆玕竿揀諫墾栞奸柬澗磵稈艱癇忓矸 杆桿 癇癎干間看刊肝幹簡姦懇           33
2        감                            勘堪瞰坎嵌憾戡柑橄疳紺邯龕玪 鑑鑒甘減感敢監鑑           23
3        강       江降講強康 杠堈岡姜橿彊慷畺疆糠絳羌腔舡薑鱇嫝跭 玒顜茳 強强 鋼鎠 岡崗剛鋼綱 鏹 襁           44
4        개                          价凱愷漑塏愾疥芥豈鎧玠 個箇 蓋盖改皆個開介慨槪蓋           25
5        거                                去巨居車擧\r距拒據渠遽鉅炬倨据祛踞鋸           18
6        건                               建乾件健巾虔楗鍵愆腱蹇騫搴湕踺建䢖 乾漧           20
7        경  京景經庚耕\r敬輕驚慶競\r竟境鏡頃傾\r硬警徑卿倞鯨坰耿炅更梗憬璟瓊擎儆俓涇莖勁逕熲冏勍\...           74
8        계                 癸季界計溪\r鷄系係戒械\r繼契桂啓階\r繫誡烓屆悸棨稽谿界堺 谿磎           31
9        고  古故固苦高\r考告枯姑庫\r孤鼓稿顧叩敲皐暠呱尻拷槁沽痼睾羔股膏苽菰藁蠱袴誥\r賈辜錮雇杲鼔...           48
10       공                              工功空共公\r孔供恭攻恐\r貢珙控拱蚣鞏龔           19
11       관                       官觀關館管\r貫慣冠寬款琯錧灌瓘梡串棺罐菅涫輨館舘 寬寛           27
12       광                             光廣鑛狂侊洸珖桄匡曠壙筐胱廣広 光炛 . 炚           22
13       교               交校橋敎郊\r較巧矯僑喬嬌膠咬嶠攪狡皎絞翹蕎蛟轎餃驕鮫姣佼噭\r憍鄗敎教           34
14       구  九口求救究\r久句舊具俱\r區驅苟拘狗\r丘懼龜構球玖矩邱銶溝購鳩軀枸仇勾咎嘔垢寇嶇柩歐毆毬...           68
15       규                             叫規糾圭奎珪揆逵窺葵槻硅竅赳閨邽嫢湀茥煃糾糺           22
16       근                               近勤根斤僅\r謹墐漌槿瑾嫤筋劤懃芹菫覲饉           19
17       기  己記起其期\r基氣技幾旣\r紀忌旗欺奇\r騎寄豈棄祈\r企畿飢器機淇琪璂棋祺錤騏麒玘杞埼崎琦...           78
18       단                        丹但單短端\r旦段壇檀斷\r團緞鍛亶彖湍簞蛋袒鄲煓㫜担           25
19       담                              談淡擔譚膽澹覃啖坍憺曇湛痰聃蕁錟潭倓啿埮炎           21
20       대                          大代待對帶\r臺貸隊垈玳袋戴擡旲岱黛㬃㬣臺坮 擡抬           24
21       도  刀到度道島\r徒圖倒都桃\r挑跳逃渡陶\r途稻導盜塗堵棹濤燾禱鍍蹈屠悼掉搗櫂淘滔睹萄覩賭韜馟...           50
22       동                       同洞童冬東\r動銅凍棟董潼垌瞳蝀憧疼胴桐朣曈彤烔橦勭同仝           27
23       람                                覽濫嵐攬欖籃纜襤藍㛦灆婪漤爁璼攬擥 .           19
24       량                              良兩量涼梁\r糧諒亮倆樑粱輛駺俍糧粮 涼凉           20
25       려                                旅麗慮勵呂侶閭黎儷廬戾櫚濾礪藜蠣驢驪曞           19
26       령                           令領嶺零靈伶玲姈昤鈴齡怜囹笭羚翎聆逞泠澪岭呤岭岺           24
27       로                          路露老勞爐魯盧鷺撈擄櫓潞瀘蘆輅鹵嚧虜璐櫨蕗潦瓐澇虜           25
28       리        里理利梨李\r吏離裏履俚莉璃俐唎浬狸痢籬罹羸釐鯉涖 犂摛裏裡 離离 俐悧\r釐厘 犂犁           41
29       린                              鄰潾璘麟吝燐藺躪鱗撛鏻獜橉粦粼䗲繗麟 鄰隣           21
..     ...                                                ...          ...
86       지  只支枝止之\r知地指志至\r紙持池誌智\r遲旨沚址祉趾祗芝摯鋕脂咫枳漬砥肢芷蜘識贄洔\r厎汦...           53
87       진  辰眞進盡振\r鎭陣陳珍震晉瑨瑱津璡秦軫塵禛診縝塡賑溱抮唇嗔搢桭榛\r殄畛疹瞋縉臻蔯袗 蓁昣枃...           63
88       차                               且次此借差車叉瑳侘嗟嵯磋箚茶蹉遮硨奲姹鹺           20
89       찬                   贊讚撰纂粲澯燦璨瓚纘鑽竄餐饌攢巑儹篡欑㜺贊賛 讚讃 儹儧\r篡簒           31
90       창                        昌唱窓倉創\r蒼暢菖昶彰敞廠倡娼愴槍漲猖瘡脹艙滄淐晿淌           26
91       척                              尺斥拓戚陟倜刺剔擲滌瘠脊蹠隻墌慼墌坧 慼慽           21
92       천                             天千川泉淺\r賤踐遷薦仟阡喘擅玔穿舛釧闡韆茜           21
93       청                               靑淸晴請廳\r聽菁鯖靑青 淸清 晴\r請           18
94       초                初草招肖超\r抄礎秒樵焦蕉楚剿哨憔梢椒炒硝礁稍苕貂酢醋醮岧釥\r俏草艸           33
95       총                                 銃總聰寵叢悤憁摠蔥冢聰聡 冢塚 總総           18
96       추                          秋追推抽醜楸樞鄒錐錘墜椎湫皺芻萩諏趨酋鎚雛騶鰌鰌鰍           25
97       치                      治致齒値置\r恥熾峙雉馳侈嗤幟梔淄痔癡緇緻蚩輜稚癡痴 稚穉           28
98       탁                               濁托濯卓度倬琸晫託擢鐸拓啄坼柝琢踔橐橐槖           20
99       태                               太泰怠殆態汰兌台胎邰笞苔跆颱鈦珆鮐脫娧迨           20
100      하                    下夏賀何河\r荷廈霞瑕蝦遐鰕呀嘏碬閜嚇赮 煆蕸㰤抲㗿廈厦 夏昰           30
101      한                          閑寒恨限韓\r漢旱汗澣瀚翰閒悍罕澖 僩嫺橌䦥扞忓邗           24
102      항                            恒巷 項\r抗航亢沆姮伉杭桁缸肛行降恒恆 姮嫦           22
103      해                         害海亥解奚\r該偕楷諧咳垓孩懈瀣蟹邂駭骸咍瑎澥祄晐海           25
104      현          現賢玄絃縣\r懸顯見峴晛泫炫玹鉉眩昡絢呟俔睍舷衒弦儇譞怰䧋\r鋗㢺琄嬛娊妶灦㭹顯顕           39
105      형                          兄刑形亨螢\r衡型邢珩泂炯瑩瀅馨熒滎灐荊鎣迥侀迥逈           24
106      혜                                 惠慧兮蕙彗譿寭憓暳蹊醯鞋譓鏸匸䚷惠恵           18
107      호  戶乎呼好虎\r號湖互胡浩\r毫豪護

In [49]:
len(df_subset)

116

This is a little more managable, we will contruct all possible 2-syllable names from these. It seems like our assumption was relatively sound, there are many common name particles present in this data set.

In [94]:
df_subset[(df_subset['hangul'] == '수') | (df_subset['hangul'] == '현')]

hangul                                              hanja  hanja_count
52       수  水手受授首\r守收誰須雖\r愁樹壽數修\r秀囚需帥殊\r隨輸獸睡遂\r垂搜洙琇銖粹穗繡隋髓袖...           92
104      현          現賢玄絃縣\r懸顯見峴晛泫炫玹鉉眩昡絢呟俔睍舷衒弦儇譞怰䧋\r鋗㢺琄嬛娊妶灦㭹顯顕           39

In [84]:
df_names = pd.DataFrame(columns=['syllable_1','syllable_2'])

In [85]:
x = 0

for i in range(len(df_subset)):
    for j in range(len(df_subset)):
        df_names.at[x, 'syllable_1'] = df_subset.at[i,'hangul']
        df_names.at[x, 'syllable_2'] = df_subset.at[j, 'hangul']
        x+=1

df_names.head(10)

syllable_1 syllable_2
0          가          가
1          가          간
2          가          감
3          가          강
4          가          개
5          가          거
6          가          건
7          가          경
8          가          계
9          가          고

In [86]:
len(df_names)

13456

In [91]:
df_names['name'] = df_names['syllable_1'] + df_names['syllable_2']
cols = df_names.columns.tolist()
cols = cols[-1:] + cols[:-1]
df_names = df_names[cols]
df_names.head(10)

name syllable_1 syllable_2
0   가가          가          가
1   가간          가          간
2   가감          가          감
3   가강          가          강
4   가개          가          개
5   가거          가          거
6   가건          가          건
7   가경          가          경
8   가계          가          계
9   가고          가          고

# Mining Name Data from the Website

In [125]:
from urllib.request import urlopen
from bs4 import BeautifulSoup

In [123]:
def convert_address(example_char):
    example_string = str(example_char.encode('utf-8'))
    example_string = example_string.replace("\\", "").replace("x","%").replace("'", "").upper()[1:len(example_string)]
    return 'https://koreanname.me/name/' + example_string

In [127]:
convert_address('수현')

'https://koreanname.me/name/%EC%88%98%ED%98%84'

Needed websites (erumy)
- http://www.erumy.com/nameAnalyze/AnalyzeMyName.aspx?name=송시안
- http://www.erumy.com/nameclub/NameResultList.aspx?name=%EC%8B%9C%EC%95%88
- http://www.erumy.com/nameclub/NameSearch.aspx?searchname=%EC%8B%9C%EC%95%88
- http://www.erumy.com/nameclub/NameResultChartYear.aspx?name=%EC%8B%9C%EC%95%88
- http://www.erumy.com/nameclub/NameResultChartLoc.aspx?name=%EC%8B%9C%EC%95%88 [line 78]

In [128]:
page = urlopen('http://www.erumy.com/nameAnalyze/AnalyzeMyName.aspx?name=%ec%86%a1%ec%84%a0%ec%9a%b0')
soup = BeautifulSoup(page, 'html.parser')
soup

<!DOCTYPE html>
<html lang="ko"><head><meta charset="utf-8"/><meta content="width=device-width, initial-scale=1, shrink-to-fit=no" name="viewport"/><meta content="이름,선호이름,신생아이름,이름통계,아기이름,작명,선호하는,아기,개명,한국이름" name="keywords"/><meta content="kimkkikki" name="author"/><meta content="https://koreanname.me/static/home.jpg" name="image"/><meta content="한국인의 이름 통계 서비스" name="description"/><meta content="한국인의 이름 통계 서비스" property="og:description"/><meta content="https://koreanname.me/static/home.jpg" property="og:image"/><meta content="a9df23b43bfd8a4b2706a8fc4ec01218e8427938" name="naver-site-verification"/><meta content="54A117C9AC83B1E74B48C93F9A779987" name="msvalidate.01"/><meta content="4DkPnpJFXVbFpfcLWF8cpznQifPoJXx9OE7sVN81MH4" name="google-site-verification"/><link href="/static/favicon.jpg" rel="icon" type="image/x-icon"/><script async="" src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script><script src="//developers.kakao.com/sdk/js/kakao.min.js"></script><script as